spec/models/competition_spec.rb

```diff
@@ -1584,4 +1584,96 @@ def change_and_check_activities(new_start_date, new_end_date)
       expect(new_competition).not_to be_valid
     end
   end
+
+  context 'auto-close registrations' do
+    let!(:auto_close_comp) { FactoryBot.create(:competition, :registration_open, auto_close_threshold: 5) }
+    let(:comp) { FactoryBot.create(:competition, :registration_open, :with_competitor_limit, competitor_limit: 3) }
+
+    it 'attempt auto close returns false when it fails' do
+      expect(auto_close_comp.attempt_auto_close!).to eq(false)
+    end
+
+    it 'attempt auto close returns true when it succeeds' do
+      FactoryBot.create_list(:registration, 4, :paid, competition: auto_close_comp)
+      reg = FactoryBot.create(:registration, competition: auto_close_comp)
+      payment = FactoryBot.build(:registration_payment, registration: reg)
+      payment.save(validate: false) # Stop the after_save hook from being triggered so that we can trigger it manually
+
+      expect(auto_close_comp.attempt_auto_close!).to eq(true)
+    end
+
+    it 'doesnt auto-close if threshold not reached' do
+      FactoryBot.create(:registration, :paid, competition: auto_close_comp)
+      expect(auto_close_comp.registration_past?).to eq(false)
+    end
+
+    it 'doesnt auto-close if threshold is null' do
+      FactoryBot.create(:registration, :paid, competition: comp)
+      expect(comp.registration_past?).to eq(false)
+    end
+
+    it 'closes registrations when the close threshold is reached' do
+      FactoryBot.create_list(:registration, 5, :paid, competition: auto_close_comp)
+      expect(auto_close_comp.registration_past?).to eq(true)
+    end
+
+    it 'closes registrations when the close threshold is exceeded' do
+      FactoryBot.create_list(:registration, 5, :paid, competition: comp)
+
+      comp.update_column(:auto_close_threshold, 5)
+      FactoryBot.create(:registration, :paid, competition: comp)
+
+      expect(comp.registration_past?).to eq(true)
+    end
+
+    it 'only auto-closes if the registrations are paid registrations' do
+      FactoryBot.create_list(:registration, 5, competition: auto_close_comp)
+      expect(auto_close_comp.registration_past?).to eq(false)
+    end
```

Comment:
I think these tests should differentiate between the hook triggering auto-close, and somebody explicitly trying to call the `auto_close!` method. Perhaps duplicate the tests, and in one batch just test whether the hook does its thing, but also have a second batch of similar/identical tests where you explicitly call the auto_close method and assume that it does/doesn't work (depending on the individual test case)

Reply:
Great point - I've separated the tests into: 
- tests against the function itself in `competition_spec`
- a test asserting that the hook does actually call `attempt_auto_close!` in `registration_spec`